### PR TITLE
bip152: check if indexes do not overflow

### PR DESCRIPTION
**Pull request by brunoerg**
When deserializing a `HeaderAndShortIds` we could check if the number of txs does not overflow 16 bits. If so, throw an error.

### CODE REVIEW

**Review by apoelstra — DISMISSED** [review]
ACK 4fcfe90b78f4aa8a34fbccef4f2e3bbe7356c57a This is a clear improvement, though I would like to see a unit test, and maybe the right approach is to make this structure private and inforce this invariant everywhere.

File: bitcoin/src/bip152.rs [hunk 1]
```diff
@@ -158,7 +158,32 @@ pub struct HeaderAndShortIds {
...
+
+impl Decodable for HeaderAndShortIds {
+    fn consensus_decode<R: BufRead + ?Sized>(r: &mut R) -> Result<Self, encode::Error> {
+        let header_short_ids = HeaderAndShortIds {
+            header: Decodable::consensus_decode(r)?,
+            nonce: Decodable::consensus_decode(r)?,
+            short_ids: Decodable::consensus_decode(r)?,
+            prefilled_txs: Decodable::consensus_decode(r)?
+        };
+        if header_short_ids.short_ids.len() + header_short_ids.prefilled_txs.len() > u16::MAX.into() {
```
Comment (Kixunil) [inline bitcoin/src/bip152.rs:170]: The `+` operation could overflow though. :D But I think it's OK, it'd have to fill whole RAM to do that so it'd be just OOM killed. (And we have the protection against this on another layer.)

File: bitcoin/src/bip152.rs [hunk 2]
```diff
@@ -158,7 +158,32 @@ pub struct HeaderAndShortIds {
...
+impl Encodable for HeaderAndShortIds {
+    fn consensus_encode<W: Write + ?Sized>(&self, w: &mut W) -> Result<usize, io::Error> {
+        let mut len = 0;
+        len += self.header.consensus_encode(w)?;
+        len += self.nonce.consensus_encode(w)?;
+        len += self.short_ids.consensus_encode(w)?;
+        len += self.prefilled_txs.consensus_encode(w)?;
+        Ok(len)
+    }
+}
```
Comment (Kixunil) [inline bitcoin/src/bip152.rs:186]: We really should have the macros separated so we wouldn't have to type this.

**Review by Kixunil — COMMENTED** [review]
The code looks good though I'd like to see some reference (link to Core code I guess) that this limit is really correct. It's getting later here, so I would rather not search it myself.

**Comment by brunoerg** [comment]
@Kixunil Bitcoin Core code for reference:
https://github.com/bitcoin/bitcoin/blob/master/src/blockencodings.h#L127

```cpp
SERIALIZE_METHODS(CBlockHeaderAndShortTxIDs, obj)
{
    READWRITE(obj.header, obj.nonce, Using<VectorFormatter<CustomUintFormatter<SHORTTXIDS_LENGTH>>>(obj.shorttxids), obj.prefilledtxn);
    if (ser_action.ForRead()) {
        if (obj.BlockTxCount() > std::numeric_limits<uint16_t>::max()) {
            throw std::ios_base::failure("indexes overflowed 16 bits");
        }
        obj.FillShortTxIDSelector();
    }
}
```

File: bitcoin/src/bip152.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (apoelstra) [reply bitcoin/src/bip152.rs:170]: I would suggest

```rust
match header_short_ids.short_ids.len().checked_add(header_short_ids.prefilled_txs.len()) {
    Some(x) if x < u16::MAX => {} // ok
    _ => return Err(...),
}
```

or similar.

[the remark this answers is not in the document]
Comment (brunoerg) [reply bitcoin/src/bip152.rs:170]: Seems good.

[the remark this answers is not in the document]
Comment (storopoli) [reply bitcoin/src/bip152.rs:170]: Yeah @apoelstra suggestion is far superior than mine.

**Comment by brunoerg** [comment]
Force-pushed addressing https://github.com/rust-bitcoin/rust-bitcoin/pull/2970#discussion_r1667108844

**Review by apoelstra — APPROVED** [review]
ACK 837f466f72b598fad67012e2449c3bb6e87de126

**Comment by tcharding** [comment]
I didn't read the c++ macros but why do we only do the block tx count check when `consensus_decode`ing? Don't we want to check this when we encode as well? Otherwise we can encode something that cannot be decoded.

**Comment by apoelstra** [comment]
@tcharding we shouldn't be able to represent something that has no consensus encoding. What we should do is make the contents of this struct private and refuse to construct things *in any way* that are invalid.

**Comment by apoelstra** [comment]
And the reason that isn't done here is that this PR is a "found an issue in the fuzzer, trying to fix it without taking too much time out of my day" PR. If we want to make big API changes then one of us needs to take it on :).

File: bitcoin/src/bip152.rs [hunk 3]
```diff
@@ -158,7 +158,32 @@ pub struct HeaderAndShortIds {
...
+        let header_short_ids = HeaderAndShortIds {
+            header: Decodable::consensus_decode(r)?,
+            nonce: Decodable::consensus_decode(r)?,
+            short_ids: Decodable::consensus_decode(r)?,
+            prefilled_txs: Decodable::consensus_decode(r)?
+        };
+        match header_short_ids.short_ids.len().checked_add(header_short_ids.prefilled_txs.len()) {
+            Some(x) if x <= u16::MAX.into() => Ok(header_short_ids),
+            _ => Err(encode::Error::ParseFailed("indexes overflowed 16 bits")),
+        }
```
Comment (Kixunil) [inline bitcoin/src/bip152.rs:173]: There's a potential inefficiency (and the same one is in Core) that it first deserializes into vec and does the check afterwards. Ideally we would at least avoid deserializing the second vec. (And we could still put a limit on the first one.)

This is not required for the PR, can be done later.

**Review by Kixunil — APPROVED** [review]
ACK 837f466f72b598fad67012e2449c3bb6e87de126

**Comment by tcharding** [comment]
> found an issue in the fuzzer, trying to fix it without taking too much time out of my day

Cool, thanks. Should I assume that context for any changes coming from @brunoerg? Also, does that mean that to get the most out of this fuzz fail (and future ones) that we should do this PR (fix the immediate problem) and also create an issue "investigate fully all implications of the fuzz fail and things that came up during the fuzz-fail-fix-PR"?

**Comment by apoelstra** [comment]
Yep. In this case I just opened https://github.com/rust-bitcoin/rust-bitcoin/issues/2984

**Comment by brunoerg** [comment]
> Cool, thanks. Should I assume that context for any changes coming from @brunoerg?

Not every change :) TBH, just did this because seemed simple but we usually open issues (if not critical, ofc) for the bugs we find with bitcoinfuzz. @apoelstra is subscribed there.

**Comment by tcharding** [comment]
Cool, thanks man.

**Pull request merged**